### PR TITLE
Overhaul PostgreSQL packaging, native extensions, and NixOS support

### PR DESCRIPTION
**Pull request by thoughtpolice**
###### Motivation for this change

I'm brief on time now, but this isn't done yet. See #38616. In short, this overhauls the namespacing and extension management of PostgreSQL packages, greatly easing the ability to test, build, and package PostgreSQL extensions properly. It also overhauls the NixOS module to support the proposed API in #38616, making it far easier to add custom extensions to your server in a correct way, no matter what version you choose.

I've been running these changes with some light testing for a few days, but I'm not done yet. I'll be amending and adding more comments to the commits and maybe squash one or two. These changes otherwise seem solid and shouldn't break anything.

Note that the NixOS module _remains backwards compatible_. Nothing should break, and you're unlikely to even notice if you don't use extensions. I consider this a good thing and ideally supporting it to 18.09 should not be terrible. However, warnings will be thrown when using the old APIs, and assertions are thrown if both are used at once in incompatible ways. (In fact you can even use a mix of the old and new APIs, in theory -- e.g. use the new `.packages` attribute with the old `.extraPlugins`, instead of the new `.plugins`.) I'm open to dropping backwards compatibility if other people think we can give it up for 18.09, though.

Finally, this also updates Nixpkgs and NixOS to **use PostgreSQL 10 and libpq 10 by default**. Although this isn't necessary, it seems like a reasonable time to make the push, since 9.6 has been the default since 17.09. PostgreSQL 11 will ideally be out around September, near the 18.09 release, but it's unlikely we'll have time to fit it in by then (depending on how things go for us, and them.)

I have also stepped up and added myself as the maintainer of the PostgreSQL module, and will be finishing the manual section.

I suggest reviewing the commits sequentially, in order, rather than at once. They are all individually small, work, and should be much easier to follow.

###### Things done

- [x] Actually finish writing the PostgreSQL section of the manual.
- [x] Look into deprecation warnings for old top-level attrs; using `builtins.trace` is annoying because e.g. it triggers on `nix search -u ...`
- [x] Look into #22653 and #38469 as well -- these should be fixed while I'm at it, IMO.
- [x] ~Bikeshed the naming of the new options? `.packages` is really close to `.package`, but I'm not sure of a better name. On the other hand, I think `.plugin` is a good name vs `.extraPlugins`~
- [x] Migrate all tests and NixOS modules to new API

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions (Fedora 25, sandboxing enabled)
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by thoughtpolice** [comment]
Please note that GitHub hates actually showing you the linear history of commits as it will appear in the history and instead always prefers showing things sorted-by-date, even when this is wrong. I rebased and re-arranged a few commits... Ugh.

**Comment by thoughtpolice** [comment]
Note to future archaeologists who will later discover my corpse here: this was the magical incantation needed to make GitHub display commits in order, as it should, by resetting the commit/author date of every individual commit to be 2-3 seconds apart.

`for x in $(git log --format=oneline master..pgsql-fixes | tac | awk '{print $1}'); do git cherry-pick $x; git commit --amend --reset-author --no-edit; sleep 2; done`

Surely this will be the solution to an untold future terror and save the world, one day.

**Comment by domenkozar** [comment]
@thoughtpolice can this go as-is and then we add further improvements?

**Comment by thoughtpolice** [comment]
Yes, I've been running these changes in a stable manner for at least a few weeks on my development machine, though I was hoping @ocharles could have given a review -- but I noticed I forgot to add him!

**Comment by thoughtpolice** [comment]
I was also hoping to add a `postgresqlWithPackages` function, so you could use postgres without NixOS, but with NIX_LIBDIR properly set up -- unfortunately this either requires some unsavory duplication of code between Nixpkgs and NixOS to retain backwards compatibility.

I believe I have a WIP patch for this, so I may add it as well.

File: nixos/modules/services/databases/postgresql.nix
Comment (danbst) [inline nixos/modules/services/databases/postgresql.nix]: Can this be made dependent on `cfg.package.version` instead of `config.system.stateVersion`?

**Review by danbst — CHANGES_REQUESTED** [review]

File: nixos/modules/services/databases/postgresql.nix
Comment (thoughtpolice) [reply nixos/modules/services/databases/postgresql.nix]: The intent is that e.g. if you were on NixOS 16.03, and you upgrade to NixOS 17.09, you should keep using PostgreSQL 9.5, because 9.6 is not on-disk compatible, and breaking your database across an upgrade isnt friendly. This is because when you installed NixOS, the (16.03) installer explicitly sets `stateVersion = "16.03"` in `configuration.nix` by default, and *that should not be changed* even beyond upgrades.

I might be misunderstanding what you mean, of course (a diff would be helpful if that's the case!)

Comment (danbst) [reply nixos/modules/services/databases/postgresql.nix]: This isn't a direct problem by itself, but rather a simplification for those who use PG older than NixOS default. When you set `services.postgresql.package` to some old version, you now have to set `services.postgresql.packages` to an older pacakgeset too, which, I think, can be detected by NixOS automatically.

**Comment by thoughtpolice** [comment]
@ocharles has given me his blessing/trust on this change in lieu of the fact he doesn't have much time to review it, but I'd love to start merging some of this in soon. I will post a message to the mailing list soon about this for any final objections.

**Comment by samueldr** [comment]
This will need to be documented in the release notes (nixos/doc/manual/release-notes/rl-1809.xml) I presume.

As a user, I really much prefer this newly presented interface for PostgreSQL stuff. :+1:

**Comment by thoughtpolice** [comment]
I'm still working on this, sorry for all the delays all. If it's any consolation, this has been running quite stable for me. :)

I've updated the release notes, issues like #22653 and #38469 should be fixed now (thanks to @basvandijk), and there is a beta of PostgreSQL 11 now available[1]. I'd like to actually get this done this week. Remaining things:

  - Finish manual (particularly backup/restore/upgrade cycle)
  - Finish migrating tests.

I expect this will probably take another few hours of focused work but I plan on landing this before the 18.09 branch.

---

[1] I am currently using this primarily to explore integration of LLVM/JIT support into the Postgres 11 betas; it's quite possible that I will *remove* the top-level attributes for postgres 11 before merging this, so people don't get any funny ideas, but in the mean time the support will be there and worked out, hopefully.

**Comment by thoughtpolice** [comment]
Some more changes:

  - There's a new `.withPackages` combinator exposed from every package set, which allows you to drop into a Nix shell with postgres+plugins inside a `buildEnv`. This is mostly useful for doing upgrades or demonstrations, but in the future could be useful for other things, too (for example: packing PostgreSQL Docker images along with extensions, using Nix).
  - PGXS previously did not respect `PREFIX` during builds, meaning every Postgres extension had to have a custom, hackily written `installPhase`. I have fixed PGXS' to obey `PREFIX`, meaning `make install PREFIX=/nix/store/...` now works for the majority of extensions, so their installPhases can be removed. This is also necessary for JIT support, which requires shipping bitcode files during `installPhase`, and would be tedious to ship correctly.
  - JIT support for PostgreSQL 11 has been extended and should be "Ready to Ship" -- I fixed most glaring problems with this feature. Please see the corresponding `postgres11` commit messages for details. The net result though is that `postgresql11.out` itself will bloat from about ~130MB to ~400MB due to needing `libLLVM.so`, but everything else should be fine. (In particular, `libpq`-based clients will not bloat at all).
  - There's now a section for doing simple, `pg_dumpall`-based upgrades in the manual. It uses the `.withPackage` technique for showing a demonstration upgrade, since you want newer `pg_dumpall` tools.

**Review by basvandijk — CHANGES_REQUESTED** [review]
Impressive work! 18.09 is getting better and better.

**Comment by basvandijk** [comment]
@thoughtpolice I added extra tests for some extensions in:

https://github.com/thoughtpolice/nixpkgs/compare/pgsql-fixes...LumiGuide:pgsql-fixes-extra-tests

I leave it up to you to decide to merge this into this your PR.

I made it so the extra tests are easily disableable which is handy if you want to debug the tests and don't want the extra complexity of certain extensions.

@nbp in the process I discovered that merging options like [`services.postgresql.plugins`](https://github.com/NixOS/nixpkgs/blob/7e937697dcad230880d541c339716c9f6f06fa76/nixos/modules/services/databases/postgresql.nix#L143) produce evaluation errors. These options should be assigned a function that takes an attribute set and returns a list with a selection of the values in the input set. Merging these options should create a new function that applies the functions of all the options to the input set and concatenate the resulting lists.

For this [I created the `selectorFunction` type](https://github.com/thoughtpolice/nixpkgs/commit/fec24c27ddf7d34b2252d2068349fdb44995b00c). It's currently defined in the postgresql module but there are at least 6 other places in nixpkgs that use these type of options:

* [`services.gitit.extraPackages`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/gitit.nix#L44)
* [`services.ihaskell.extraPackages`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/ihaskell.nix#L22)
* [`services.octoprint.plugins`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/octoprint.nix#L68)
* [`services.xserver.windowManager.xmonad.extraPackages`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/window-managers/xmonad.nix#L30)
* [`services.xserver.windowManager.exwm.extraPackages`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/window-managers/exwm.nix#L27)
* [`services.hoogle.packages`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/development/hoogle.nix#L27)

What do you think about moving the `selectorFunction` type to `lib/types.nix` and patching these module to use it?

**Review by basvandijk — APPROVED** [review]

**Comment by basvandijk** [comment]
What prevents us from cherry-picking all of this on `release-18.09` before the release publication?

**Comment by basvandijk** [comment]
@thoughtpolice I merged `master` into `pgsql-fixes` in [LumiGuide:pgsql-fixes-master-merge](https://github.com/NixOS/nixpkgs/compare/master...LumiGuide:pgsql-fixes-master-merge) and fixed the conflicts. `nix-build nixos/tests/postgresql.nix` succeeds.

However, I realize it might actually be better to rebase `pgsql-fixes` on `master` to make it easier to cherry-pick the commits on `release-18.09` later on if we decide that we want to do that. I'll do that later this evening.

**Comment by basvandijk** [comment]
I rebased on `master` to fix the eval errors.

@GrahamcOfBorg test postgresql

**Comment by basvandijk** [comment]
Cherry-pick of this PR on `release-18.09`: https://github.com/NixOS/nixpkgs/compare/release-18.09...LumiGuide:pgsql-fixes-release-18.09. 

The postgresql tests pass. The manual incl. release notes build successfully and look correct.

**Comment by basvandijk** [comment]
I missed that we still have an unchecked TODO item: "Migrate all tests and NixOS modules to new API". Indeed modules like [`hydron`](https://github.com/NixOS/nixpkgs/blob/9a88df84997d0c6ba57132156ac5f22496535c1d/nixos/modules/services/web-servers/hydron.nix#L108) haven't been migrated yet.

Every module that revers to `config.services.postgresql.package` needs to replace that with:
```nix
if (config.services.postgresql.package != null) 
then config.services.postgresql.package 
else config.services.postgresql.packages.postgresql
```
This is quite cumbersome and error-prone to repeat everywhere.

@thoughtpolice may I propose that we turn the local variable [`postgresqlPackage`](https://github.com/NixOS/nixpkgs/blob/9a88df84997d0c6ba57132156ac5f22496535c1d/nixos/modules/services/databases/postgresql.nix#L18) into a [`readOnly`](https://github.com/NixOS/nixpkgs/blob/9a88df84997d0c6ba57132156ac5f22496535c1d/lib/options.nix#L22) option of the `postgresql` module? Then other modules can refer to that option abstracting over the if..then..else.

**Comment by basvandijk** [comment]
@thoughtpolice I've now introduced the read-only and invisible `config.services.postgresql.postgresqlPackage` option in the last commit of https://github.com/NixOS/nixpkgs/compare/master...LumiGuide:pgsql-fixes-postgresqlPackage (https://github.com/NixOS/nixpkgs/commit/a1d18b73a3388f6a2f0275adb8bf337f1aa2336d). That commit also migrates all remaining modules to the new scheme. What do you say? Merge it here?

**Comment by vcunat** [comment]
It's better, unless there's something that's urgent (e.g. security fixes).  You just merge to the staging branch.  

Everything is manual.  Someone chooses some version of staging to stabilize &ndash; merges it to `staging-next` to stop inflow of other large changes, then repeatedly regressions are fixed (if any) and then it's merged to master.  We recently added that intermediate step https://github.com/NixOS/rfcs/pull/26 (not merged yet :)

**Comment by basvandijk** [comment]
@vcunat thanks for the excellent explanation and RFC. Then I think I will be merging this into `staging` soon.

@vcunat as 18.09 release manager, is there a chance that [this can still be cherry-picked](https://github.com/NixOS/nixpkgs/compare/release-18.09...LumiGuide:pgsql-fixes-release-18.09) in `release-18.09` once this is merged in `master`? Or is that out of the question?

@GrahamcOfBorg test postgresql.postgresql10  postgresql.postgresql11  postgresql.postgresql93  postgresql.postgresql94  postgresql.postgresql95  postgresql.postgresql96

**Comment by thoughtpolice** [comment]
Sorry about slacking on this folks, I was sick during the beginning of September which meant I missed the fork deadline. :( I'm willing to shepard these changes in still, providing we have enough time.

I'll review the rest of the info in this thread shortly.

**Comment by basvandijk** [comment]
@thoughtpolice cool to have you back. Let me know how I can help.

**Comment by thoughtpolice** [comment]
So I think the immediate thing we should do is merge this to `staging` ASAP, in order to go ahead and get this in the pipeline to hit `master`. What I will probably do is rebase things again in order to get them in a clean state on top of `staging`, since it's still active right now, and it will trickle in through `staging-next`. I think most of the downstream consumers (the 1,200 packages) are mostly just due to PostgreSQL 10 becoming the default, requiring all `libpq` clients to be rebuilt.

Doing this I think will mostly just require fixing up the remaining tests and hitting The Big Button. (Unfortunately there was also a regression in the PostgreSQL 11 JIT support, meaning the closure size is bloated again, but I think I can fix it easily.)

The NixOS 18.09 question is trickier. I have asked @samueldr about an exception on IRC, and I would also like @vcunat's opinion, so I await their judgement. @basvandijk, it looks like you're already doing this in your LumiGuide fork[1] -- so if we get an exception, I think we can use that as the basis for a cherry pick into `staging-18.09` nicely as well, which will hopefully get brought in one final time before the release.

In the event this is barred, I'd suggest we can keep maintaining an independent fork that we merge with `release-18.09` regularly, for people who want it, although it would be unfortunate (but understandable).

Realistically, my biggest concern about merging this to 18.09 is if the API needs to change again for 19.03, due to feedback from users. That would cause some annoying churn between two releases. *But*, I think even if I had merged this a month ago, we'd still be in the same place -- the vast majority of users would not be updating or testing this new API until after the release _anyway_. Meaning the end result would have been the same: we probably would only have reliable feedback well after "The point of no return". I think that's probably just part of the game...

Alternatively, if this hit 18.09 --  we could take any feedback from this new API and roll it into 19.09, in order to ensure we have adequate time for cleanup of prior code, and any future amendments, leaving the gap of 19.03 for one release. This would also be livable, in my opinion.

@vcunat, @samueldr -- please let us know what you think, and @basvandijk and I can certainly handle it.

---

[1] Just curious, are you already using this work at LumiGuide? I've been using it successfully for several months on my server, so I'm curious.

**Comment by basvandijk** [comment]
@thoughtpolice sounds like a plan.

> Just curious, are you already using this work at LumiGuide?

No, not yet but today I started the migration to 18.09 so hopefully I will be able to use it soon.

@jchildren I see you're giving a talk at [NixCon 2018](https://nixcon2018.org/#program) about building and testing PostgreSQL extensions with Nix. So you're probably interested in this work as well.

**Comment by thoughtpolice** [comment]
Wait, @basvandijk -- how did you push to *my* fork exactly? I see this change on https://github.com/thoughtpolice/nixpkgs/tree/pgsql-fixes, but how can you push there? I don't have any collaborators, so this seems really weird!

**Comment by samueldr** [comment]
> The NixOS 18.09 question is trickier. I have asked @samueldr about an exception on IRC, and I would also like @vcunat's opinion, so I await their judgement. 

For the record [the conversation we had](https://logs.nix.samueldr.com/nixos-dev/2018-09-11#1536690638-1536691453;).

I'm not too hot on changing the semantics of stateVersion == 18.09 so close to the release, but not totally entirely against it. Though, combined with the idea that this is a huge change post-freeze, it's a bit spooky IMHO. My main thinking was seeing how anybody else concerned would react to the backport news, and seeing only agreement I would go for it, *since we're still at the beta stage*. Though, the roadmap makes me think it'll be late beta, and possibly hard to go for it.

Anyway, let's see other opinions on backporting post feature-freeze.

**Comment by danbst** [comment]
I guess those who rely on Postgresql will appreciate the fix of packaging situation, even if it requires postfixups in release branch.

**Comment by basvandijk** [comment]
Since this PR also touches the vital hydra module:
@GrahamcOfBorg test hydra

**Comment by basvandijk** [comment]
The hydra test failure is due to the postgresql systemd service taking more than 2m to start up causing a timeout. Probably because ofborg is under heavy load. The test succeeds on my system.

**Comment by basvandijk** [comment]
@thoughtpolice is this already merged with `staging`? (I don't have my laptop with me so I it's hard to check myself).

**Comment by thoughtpolice** [comment]
No, I have some WIP branch to merge/rebase it on staging  myself; meant to get to it this weekend but was distracted the whole time. Should come later tonight.

**Comment by domenkozar** [comment]
I've pushed 29660a208552a1e32f872333d6eb52e13226effa (and to 18.09) using same patch as in this PR, as it was really annoying me :)

**Comment by basvandijk** [comment]
@thoughtpolice to fix the merge conflicts I've rebased `pgsql-fixes` on `master` again at my [`pgsql-fixes-master-rebase-2018-09-28`](https://github.com/NixOS/nixpkgs/compare/master...LumiGuide:pgsql-fixes-master-rebase-2018-09-28) branch. Shall I force-push that to this PR or do you want to take the lead with this?  Note that `staging` merges without conflicts into `pgsql-fixes-master-rebase-2018-09-28`.

**Comment by danbst** [comment]
API to override postgresqlPackages isn't exposed.

I can make an override for `postgresql`:
```
       pg11 = (import lumiPkgs { }).postgresql11Packages.postgresql.overrideAttrs (super: rec {
         name = "postgresql-${version}";
         version = "11beta4";
         src = pkgs.fetchurl {
           url = "mirror://postgresql/source/v${version}/${name}.tar.bz2";
           sha256 = "0qxlfh1a7bhhamrbs3msk71pny7jxx0c0fs26zlmp7jjn138zqi1";
         };
       });
```
but can't convert it into `postgresql11beta4Packages` - `makePackageSet` isn't exposed. And extensions won't be built against my patched postgresql.

**Comment by danbst** [comment]
They way, how extensions are filtered by PG version, isn't user friendly. For example, `pg_repack` is said to be compatible only with PG < 11, so it is absent from `postgresql11Packages`. This results into `attribute pg_repack not found` error, which wasn't obvious "why" (I know it should be there, because I've packaged it!).

Maybe show some warning "pg_repack extension is available only for PG < 11" too?

**Comment by danbst** [comment]
@thoughtpolice @basvandijk PG11 is released, let's merge this?

**Comment by thoughtpolice** [comment]
Unfortunately #48664 and related changes have caused a bit of a rebasing nightmare in the mean time due to attribute-related changes, so there's a cascade of fixes I need to do. This will also complicate the merge to staging, I think. I plan on spending the rest of the day trying to work this out but it'll probably take some time.

In the mean time I'm going to add myself as the owner of Postgres-related changes under Nixpkgs to make myself a blocking reviewer, so this doesn't happen in the future and I have time to review the changes people make, e.g. this PR already has PG 11 so the prior one was a bit redundant in the long run. (Not that anyone could be blamed about this, really, except for myself perhaps for not doing this sooner, and not keeping up with the changes people do submit!)

**Comment by thoughtpolice** [comment]
Update: I made a lot of good progress yesterday but I basically slowly rewrote and compressed several of the patches from the original branch to be fewer in number, and easier to grok. I hope later today I'll be ready to overwrite this branch with a new version composed on `staging` instead and just slap the merge button.

**Comment by thoughtpolice** [comment]
4569ee7d74d505f0a9b3fb8754951393e8be7e6f also fixes a recent-ish plv8 build failure.

**Comment by domenkozar** [comment]
I think this PR is a good example why we encourage contributions to be split up into small pieces.

nixpkgs moves *very* fast and PRs that hang around half a year, accumulating changes across 50+ files are too big pain to review and merge in timely manner (given that we work async).

**Comment by thoughtpolice** [comment]
I've mostly been gone for vacation and thanksgiving, FWIW. I just didn't get around to posting my staging delta, which is now substantially smaller than this PR (in number of commits and changes.) In fact, prior to my leave, I *was* piece-wise upstreaming several pieces directly into staging -- hence why I was doing several regular merges (master -> staging) a few weeks ago, to prep things (to make sure `master` wouldn't needlessly conflict, later on). So my changes have gotten smaller and more pointed, I believe.

The real issue is that multiple changes here cause mass-rebuilds and reordering them or splitting them up doesn't really change that fact, so you might as well change a bunch of it at once, if you ask me. From my count, my new branch has roughly 5 major commits that will rebuild the vast majority of all PostgreSQL builds and their downstream consumers. They each imply large, delicate changes so keeping them distinct is proper. I consider Hydra and Borg somewhat sacred resources so I try not to stress them too hard without a good reason. (It also improves my cycle time to do this anyway: my local machine is extremely fast and can build hundreds of complex expressions on the order of minutes, giving me much better turnaround than borg, at least until I hit the 'long tail' of every-downstream consumer, where I just can't have enough cores.) This isn't really a case of me being "right" though, it's just a justification for why I've acted so.

That said I understand the frustration; the reality is that my time is simply much slimmer these days (for a number of reasons, ranging from being busy to personal health) and I'm not actively paid to do maintenance like this, but at the time I needed it, I really wanted it. It probably does not help that I'm a fairly established maintainer -- I can willingly deal with occasional rebases and whatnot for my personal work and I have a good idea of what to do if things go wrong. But many ordinary users (as well as other devs) who want these changes now probably don't want to do that, understandably.

FWIW, my current delta looks like this for `staging`:

```
93e05631b04390a94ecbd7ebd44d8e5c93b81d7a postgresqlPackages.postgis: remove outdated comment
230049105532faf9e16954fde92b8859ced5055a postgresqlPackages: add new withPackages combinator, fix man page
68216f892e08d11af411b5907a95c813e494135c postgresql_11: use llvmPackages_7
eaac5719c672a8743141692e0f14945202108504 postgresql11: implement and enable JIT support
80e9a93579adabfbc63cb1310d4759fd42f1ccfc postgresqlPackages: make PGXS install phase Nix-compatible
6590a941ea9190af871ee51d192e86aec6bdf226 postgresql: always create $out/bin
bc6996beb431b4b925bd4cbad25145c85238000c nixos/tests: migrate to new postgresql api
9802c1eb06684d0ff3a16821bc82e513fc04dc25 nixos/tests: migrate postgis test to new postgres API
2dae770378b17efa617554f9a750b6a14a7694fb nixos/postgresql: give the plugins option the selectorFunction type
f9c33be8329f345f3ceb5dce322e83a663094474 nixos/postgresql: add new nixos module API
f195107caad0a26c533a715e45a249a375ee3038 postgresql: enable systemd service notification on 9.6+
9b86b39cc8792e03d3779992484d733da47663a3 postgresqlPackages: add versionCheck passthru for checking server compatibility
74ef5728a645fdda3b4f3b09e212d53c6e144593 nixpkgs: migrate postgis to new postgresqlPackages infrastructure
d0a41fde11b8706268a5f6692a13efda158f5952 nixpkgs: implement new postgresqlPackages infrastructure
c876b993145acdcd7e605faf9744690292b3a910 nixpkgs: reorganize the postgresql extensions
3afeb2d7895512754f68ef6335846346ffa7fd60 nixpkgs: move postgresql patches into a common directory
```

I can of course go ahead and merge *that* to staging. In fact it should work quite fine, I think. So perhaps I should finally start cycling `borg` once I get these changes posted. (Some of it is arguably out of scope, such as the LLVM work; this could also be removed, I suppose, but again it's going to cause a big rebuild anyway.)

This is quite minimal and isolated in my opinion; close to what you'd need to hit all the high points. The remaining points are:

  - Test migration isn't actually done
  - Manual needs to be rewritten (I had most of this but the merge conflicts were too great to work out, due to a pass of automatic formatting, so this is "done" but I have to reformat, prepare it, etc).

Other than that I do think it's almost ready. Realistically though I'm just very busy and testing all these combinations takes a bit of time with each iteration and packaging them up nicely; I think fundamentally the scope of the changes is a little large no matter how you slice it.

**Comment by danbst** [comment]
@thoughtpolice probably we can split this PR somehow, to speedup merge? Don't want to miss pg11 in 19.03

**Comment by danbst** [comment]
I've started split. My plan:
- [x] `postgresqlPackages`, and make them extensible! https://github.com/NixOS/nixpkgs/pull/54319
- [ ] fix stuff, so all should work with pg11
- [ ] change default to pg11. This is minimal plan for 19.03. Eveything else can go step by step
- [ ] enable JIT in pg11
- [ ] make aliases low-prio, so they don't appear in nix search. Instead, force postgresqlPackages to be present there
- [ ] rething NixOS module options (in regard to plugins)
- [ ] everything else

**Comment by thoughtpolice** [comment]
I had actually started up on this again last week, but I've basically been in-and-out of the house quite a lot. I also split up some things that I think can go into staging quicker now, so I'll try to take a look at #54319 when I get a chance and we'll see if we can meet in the middle.

**Comment by danbst** [comment]
@thoughtpolice this PR is heavily outdated now, I'll close it. Most of this stuff is already ported. The only non-ported things (AFAIK):
- JIT support
- PG version constraints for extensions

Feel free to open a PRs for that.

**Pull request closed without merging**
